File: 08-streaming/8.1-over-tcp/src/main/scala/x/OverTCP.scala
package x


import org.apache.spark.SparkConf
import org.apache.spark.streaming.StreamingContext
import org.apache.spark.streaming.Seconds
import org.apache.spark.storage.StorageLevel

/*
$  ~/spark/bin/spark-submit  --master local[2]   --driver-class-path logging/  --class x.OverTCP target/scala-2.11/over-tcp_2.11-1.0.jar
 */


object OverTCP {
  def main(args: Array[String]) {

    val sparkConf = new SparkConf().setAppName("BlkIPOverTCP")

    // TODO  1 : define window duration for 5 seconds
    val ssc = new StreamingContext(sparkConf, Seconds(5))

    // TODO-2: listen on port 10000, only cache it local memory
    val lines = ssc.socketTextStream("localhost", 10000, StorageLevel.MEMORY_ONLY)
    lines.print

    // TODO-2 : filter lines that contains 'x'
    val blocked = lines.filter(line => line.contains("x"))
    val blocked2 = blocked.map("##FILTERED:" + _) // better print
    blocked2.print

    // TODO-3  : Save both RDDs (and uncomment this block)
    // blocked.saveAsTextFiles("out")

    // BONUS-LAB
    // hint  : data  format:
    // timestamp, ip, userid, action, domain, campaign, cost, sessionid
    // 1420070400000,ip_1,user_5,clicked,facebook.com,campaign_6,139,session_98
    // hint : separator is comma
    // hint : ip address is second element : index (1)

    // extract lines with more than one column
    val blocked3 = blocked.filter(_.split(",").size > 1)
    val blockedIPs = blocked3.map(line => line.split(",")(1))
    val blockedIPs2 = blockedIPs.map("BLOCKED-IP:" + _) // better print
    blockedIPs2.print()

    // kick off the processing
    ssc.start()
    ssc.awaitTermination()
  }
}

File: 08-streaming/8.1-over-tcp/python/over-tcp.py
from pyspark import SparkContext
from pyspark.streaming import StreamingContext

# Create a local StreamingContext with two working thread and batch interval of 1 second
sc = SparkContext("local[2]", "OverTCP")

## TODO-1 : Create a streaming context with 5 second
## batch interval
## Hint : 5
ssc = StreamingContext(sc, 5)

## TODO-2 Create a DStream on localhost:10000
lines = ssc.socketTextStream("localhost", 10000)
lines.pprint()


## TODO-3 : filter for lines that has 'blocked'
blocked= lines.filter(lambda line : "blocked" in line)
blocked.pprint()

## TODO-4 : Save both RDDs 
blocked.saveAsTextFiles("out-blocked")


ssc.start()             # Start the computation
ssc.awaitTermination()  # Wait for the computation to terminate
